lib/nmatrix/lapacke.rb

```diff
@@ -232,4 +232,111 @@ def solve(b, opts = {})
       raise(ArgumentError, "#{opts[:form]} is not a valid form option")
     end
   end
+
+  #
+  # call-seq:
+  #     geqrf! -> shape.min x 1 NMatrix 
+  #
+  # QR factorization of a general M-by-N matrix +A+. 
+  #
+  # The QR factorization is A = QR, where Q is orthogonal and R is Upper Triangular
+  # +A+ is overwritten with the elements of R and Q with Q being represented by the 
+  # elements below A's diagonal and an array of scalar factors in the output NMatrix. 
+  #
+  # The matrix Q is represented as a product of elementary reflectors
+  #     Q = H(1) H(2) . . . H(k), where k = min(m,n).
+  #
+  # Each H(i) has the form
+  #
+  #     H(i) = I - tau * v * v'
+  #
+  # http://www.netlib.org/lapack/explore-html/d3/d69/dgeqrf_8f.html
+  # 
+  # Only works for dense matrices.
+  #
+  # * *Returns* :
+  #   - Vector TAU. Q and R are stored in A. Q is represented by TAU and A
+  # * *Raises* :
+  #   - +StorageTypeError+ -> LAPACK functions only work on dense matrices.
+  #
+  def geqrf!
+    raise(StorageTypeError, "LAPACK functions only work on dense matrices") unless self.dense?
+    
+    tau = NMatrix.new([self.shape.min,1], dtype: self.dtype)
+    NMatrix::LAPACK::lapacke_geqrf(:row, self.shape[0], self.shape[1], self, self.shape[1], tau)
+    
+    tau
+  end
+  
+  #
+  # call-seq:
+  #     ormqr(tau) -> NMatrix
+  #     ormqr(tau, :left, false, C) -> NMAtrix
+  #
+  # Returns Q as the product Q * C after a call to geqrf! used in QR factorization. 
+  #  
+  # Only works on float types, use unmqr for complex types.
+  #
+  # == Arguments
+  #
+  # * +tau+ - vector containing scalar factors of elementary reflectors
+  # * +side+ - direction of multiplication [:left, :right]
+  # * +transpose+ - Apply Q with or without transpose [false, :transpose] 
+  # * +multiplicand+ - Multiplicand of Q * C, no argument assumes C = I
```

Comment:
Q*C or C*Q depending on the arguments. Also, you should mention that this argument will be modified.